### PR TITLE
Issue/136

### PR DESCRIPTION
**Pull request by fulmicoton**
This is quite a bit of a redesign of the indexer, mostly for performance reasons.

- The Hashmap is now shared for all of the fields
- The hash table now uses quadratic probing. Even at low capacity, linear probing was very pathological when there was a lot of distinct terms.
- We do not call `.close_doc()` for all of the terms recorder. When there was a lot of terms, this would create a huge raise in the size of the heap, which would exceed its limit and force reallocation.
- We avoid copying terms during postings serialization

### CODE REVIEW

File: src/datastruct/stacker/hashmap.rs [hunk 1]
```diff
@@ -57,34 +53,58 @@ pub enum Entry {
...
-    #[inline]
-    fn bucket(&self, key: &[u8]) -> usize {
-        let hash: u64 = djb2(key);
-        (hash as usize) & self.mask
+    fn probe(&self, key: &[u8]) -> QuadraticProbing {
+        QuadraticProbing::compute(key, self.mask)
+    }
+
+    pub fn is_saturated(&self) -> bool {
+        self.table.len() < self.occupied.len() * 5
```
Comment (arthurprs) [inline src/datastruct/stacker/hashmap.rs:106]: You are enforcing a max load factor of 20%. That's way too aggressive.

Comment (fulmicoton) [reply src/datastruct/stacker/hashmap.rs:106]: You are right. I'll test different ratio.

**Comment by fulmicoton** [comment]
> Is this strategy general? If you have high cardinality data it can potentially generate much more segments than expected.

You mean not resizing the hash table ? Yes it is general.

> If you have high cardinality data it can potentially generate much more segments than expected.

It is not as bad as you think it is.
So imagine we allocate half of the budget to hashmaps, and half to the heap.
By definition, by not knowing what should have been the perfect balance, we wasted **at most** half of the memory allocated.

Now what about the more sophisticated solution. Surely we do not want to rehash the table, so we'll start storing part of the hashes... It's a lot of complexity.

**Comment by arthurprs** [comment]
I guess the key thing then is splitting the memory between the memory and index. If that's done in advance or tracked as the indexing progresses, etc..

> Now what about the more sophisticated solution. Surely we do not want to rehash the table, so we'll start storing the hashes... That suddenly makes every entry twice as big.

Well, it's not as clear cut as that. You are already storing somewhat equivalent metadata by keeping track of buckets with entries. I assume it's to make iteration faster due to the sparse table. There's tradeoffs here and I understand the one you choose.

File: src/indexer/merger.rs [hunk 2]
```diff
@@ -201,6 +201,8 @@ impl IndexMerger {
             }
             merged_doc_id_map.push(segment_local_map);
         }
+
+        let mut field = Field(u32::max_value());
```
Comment (KodrAus) [inline src/indexer/merger.rs]: What do we use `u32::max_value` for?

File: src/indexer/segment_writer.rs [hunk 3]
```diff
@@ -129,6 +94,16 @@ impl<'a> SegmentWriter<'a> {
 		self.heap.num_free_bytes() <= MARGIN_IN_BYTES
 	}
 	
+	
+	/// Return true if the term dictionary hashmap is reaching capacity.
+	/// It is one of the condition that triggers a `SegmentWriter` to
+	/// be finalized.
+	#[doc(hidden)]
+	pub fn is_termdic_saturated(&self,) -> bool {
```
Comment (KodrAus) [inline src/indexer/segment_writer.rs]: Should this be public?

File: src/indexer/segment_writer.rs [hunk 4]
```diff
@@ -214,14 +188,16 @@ impl<'a> SegmentWriter<'a> {
...
-			try!(per_field_postings_writer.serialize(serializer.get_postings_serializer(), heap));
-		}
+fn write<'a>(
+		multifield_postings: &MultiFieldPostingsWriter,
+		fast_field_writers: &FastFieldsWriter,
+		fieldnorms_writer: &FastFieldsWriter,
+	  	mut serializer: SegmentSerializer) -> Result<()> {
+		
+		try!(multifield_postings.serialize(serializer.get_postings_serializer()));
+		// for per_field_postings_writer in per_field_postings_writers {
```
Comment (KodrAus) [inline src/indexer/segment_writer.rs]: Do you still need this commented out code?

File: src/schema/term.rs [hunk 5]
```diff
@@ -15,6 +15,12 @@ const INT_TERM_LEN: usize = 4 + 8;
 #[derive(Clone, PartialEq, PartialOrd, Ord, Eq, Hash)]
 pub struct Term(Vec<u8>);
 
+/// Extract `field` from Term.
+#[doc(hidden)]
```
Comment (KodrAus) [inline src/schema/term.rs]: Do missing doc checks not understand `pub(restricted)`? If this is only public to the crate it shouldn't emit a missing doc warning.

**Review by KodrAus — COMMENTED** [review]
I don't fully grok the change so will leave that discussion to the experts :) I'm happy when you are.

Just left a few nits.

File: src/schema/term.rs (hunk 5 above)
Comment (fulmicoton) [reply src/schema/term.rs]: will clean up!

File: src/indexer/segment_writer.rs (hunk 4 above)
Comment (fulmicoton) [reply src/indexer/segment_writer.rs]: will clean up!

File: src/indexer/segment_writer.rs (hunk 3 above)
Comment (fulmicoton) [reply src/indexer/segment_writer.rs]: will clean up!

File: src/indexer/merger.rs (hunk 2 above)
Comment (fulmicoton) [reply src/indexer/merger.rs]: I'll change it for an Option<>

Right now `u32::max_value()` is here to make sure the first field is added.

**Pull request merged**